src/Http/Wolverine.Http.Marten/CompiledQueryWriterPolicy.cs

```diff
@@ -14,82 +15,65 @@ public class CompiledQueryWriterPolicy : IResourceWriterPolicy
 {
     public bool TryApply(HttpChain chain)
     {
-        bool ImplementsGenericInterface(Type type, Type assigned)
+        bool ImplementsGenericInterface(Type type, Type assigned, [NotNullWhen(true)] out Type[]? types)
         {
-            return type.GetInterfaces().Any(x =>
+            types = null;
+            var result = type.GetInterfaces().FirstOrDefault(x =>
                 x.IsGenericType && x.GetGenericTypeDefinition() == assigned);
+            if (result is null) return false;
+            types = result.GetGenericArguments();
+            return true;
         }
 
         var result = chain.Method.Creates.FirstOrDefault();
         if (result is null) return false;
-        if (ImplementsGenericInterface(result.VariableType, typeof(ICompiledListQuery<,>)))
-        {
-            chain.Postprocessors.Add(new MartenWriteArrayCodeFrame(result));
-            return true;
-        }
-
-        if (ImplementsGenericInterface(result.VariableType, typeof(ICompiledQuery<,>)))
+        // Are we dealing with a compiled query or not?
+        if (ImplementsGenericInterface(result.VariableType, typeof(ICompiledQuery<,>), out var arguments))
         {
-            chain.Postprocessors.Add(new MartenWriteJsonCodeFrame(result));
-            return true;
-        }
-
-        return false;
-    }
-}
-
-public class MartenWriteJsonCodeFrame : AsyncFrame
-{
-    private Variable? _documentSession;
-    private Variable? _httpContext;
-    private readonly Variable _compiledQuery;
-
-    public MartenWriteJsonCodeFrame(Variable variable)
-    {
-        _compiledQuery = variable;
-    }
+            MethodInfo methodToCall;
```

Comment:
This kind of thing is easier if you subclass MethodCall like this guy: https://github.com/JasperFx/wolverine/blob/main/src/Persistence/Wolverine.Marten/Codegen/RegisterEventsFrame.cs, then use this helper to create the frame:

https://github.com/JasperFx/wolverine/blob/feecee5d019adac3f37e89c95b243c4835db15da/src/Persistence/Wolverine.Marten/AggregateHandlerAttribute.cs#L127